File: 1 SEMESTRE/EXERCICIOS DE LISTA, PILHA E FILA/Lista de Pilhas/Exercicios em C/Exercicio2.c
/*Mesmo exercício 2 de java, porém em C.*/

//Struct para a Lista. 
typedef struct Lista {
    //Estrutura da Lista. 
    struct CelulaLista *inicio; 
    struct CelulaLista *fim; 
} Lista; 

//Struct para a Célula da Lista, cada nó. 
typedef struct CelulaLista {
    //Estrutura que cada nó da lista possui. 
    struct CelulaPilha *topo; 
    struct CelulaLista *prox; 
} CelulaLista; 

//Struct para a Célula da Pilha, cada nó 
typedef struct CelulaPilha {
    //Estrutura que cada nó da pilha possui. 
    int elemento; 
    struct CelulaPilha *prox; 
} CelulaPilha; 

CelulaLista *menorPilha(Lista *lista) {
    //Variável para armazenar a lista 
    CelulaLista *menor = lista->inicio; 
    int menorQntd = 0; 

    //Percorrer a primeira pilha para ter o contador como referência. 
    for (CelulaPilha *k = lista->inicio->topo; k != NULL; k = k->prox) {
        menorQntd++; //Saber quantidade de elementos da primeira pilha. 
    }

    //Percorrer a lista a partir da segunda pilha. 
    for (CelulaLista *i = lista->inicio->prox; i != NULL; i = i->prox) {
        //Contador atual para saber a quantidade de elementos de cada pilha. 
        int qntdAtual = 0; 

        //Percorrer a segunda pilha. 
        for (CelulaPilha *j = i->topo; j != NULL; j = j->prox) {
            qntdAtual++; 
        }

        //Verificar
        if (qntdAtual < menorQntd) {
            menorQntd = qntdAtual; 
            menor = i; 
        }
    }

    //Retornar o ponteiro da lista com a menor pilha. 
    return menor; 
}

File: 1 SEMESTRE/EXERCICIOS DE LISTA, PILHA E FILA/Lista de Pilhas/Exercicios em Java/Exercicio2.java
/*Conside uma lista de pilhas, dessa vez, em vez de retornar a maior pilha, o método agora vai retornar a menor pilha, ou seja, a pilha
que possui a menor quantidade de elementos. */

//Classe para Lista, que vai ter um ponteiro para inicio e outro para o fim da lista. 
class Lista {
    CelulaLista inicio; 
    CelulaLista fim; 
}

//Classe para a célula da lista, que vai cada nó vai ter um topo e da pilha e um prox, apontando pro próximo elemento da lista. 
class CelulaLista {
    CelulaPilha topo; 
    CelulaLista prox; 
}

//Classe para a célula da pilha, que vai ter um elemento, e um prox, apontando para o próximo elemento da pilha. 
class CelulaPilha {
    int elemento; 
    CelulaPilha prox; 
}

//Método para procurar e retornar a menor pilha. 
CelulaLista menorPilha() {
    //Inicializar variável 
    CelulaLista menor = inicio; //Vai ter a célula da lista (nó) que possui a menor quantidade. 
    int menorQntd = 0; //Variável para salvar a quantidade de cada pilha. 

    //Percorrer a primeira pilha para ter base para comparar com as outras. 
    for (CelulaPilha k = inicio.topo; k != null; k = k.prox) {
        menorQntd++; //Incrementar o contador. 
    }

    //Percorrer primeiramente a lista, a partir da segunda pilha, já que a primeira já foi verificada
    for (CelulaLista i = inicio.prox; i != null; i = i.prox) {
        //Variável para contar a quantidade atual de elementos de cada pilha. 
        int qntdAtual = 0; 
        
        //Percorrer os elementos da pilha até que seja diferente de null. 
        for (CelulaPilha j = i.topo; j != null; j = j.prox) {
            qntdAtual++; //Incrementar o contador
        }

        //Verificar se a quantidade atual é menor que a menor quantidade. 
        if (qntdAtual < menorQntd) {
            menorQntd = qntdAtual; 
            menor = i; 
        }
    }

    //Retornar a célula da lista com a menor quantidade de elementos de cada pilha. 
    return menor; 
}